tests/utils.py

```diff
@@ -21,6 +22,102 @@
+        self.__class__.__name__ = model.__name__
+
+    def fit(self, X, y):
+        init_params = self.params.get("init", {})
+        self.fit_intercept_ = init_params.pop("fit_intercept", False)
+        if self.fit_intercept_:
+            X = sm.add_constant(X)
+        est = self.model(y, X, **init_params)
+        if "fit_regularized" in self.params:
+            self.fitted_model_ = est.fit_regularized(
+                **self.params["fit_regularized"])
+        else:
+            self.fitted_model_ = est.fit(**self.params.get("fit", {}))
+        self.__class__.__name__ = type(self.fitted_model_).__name__
+        return self.fitted_model_
+
+    def predict(self, X):
+        if self.fit_intercept_:
+            X = sm.add_constant(X)
+        return self.fitted_model_.predict(X)
+
+
+class ModelTrainer:
+
+    _class_instances = {}
+
+    def __init__(self, dataset_name, test_fraction):
+        self.dataset_name = dataset_name
+        self.test_fraction = test_fraction
+        np.random.seed(seed=7)
+        if dataset_name == "boston":
+            self.__name__ = "train_model_regression"
+            dataset = datasets.load_boston()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "iris":
+            self.__name__ = "train_model_classification"
+            dataset = datasets.load_iris()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "breast_cancer":
+            self.__name__ = "train_model_classification_binary"
+            dataset = datasets.load_breast_cancer()
+            self.X, self.y = shuffle(
+                dataset.data, dataset.target, random_state=13)
+        elif dataset_name == "regression_rnd":
+            self.__name__ = "train_model_regression_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.random(size=(N, 1))
+        elif dataset_name == "classification_rnd":
+            self.__name__ = "train_model_classification_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.randint(3, size=(N,))
+        elif dataset_name == "classification_binary_rnd":
+            self.__name__ = "train_model_classification_binary_random_data"
+            N = 1000
+            self.X = np.random.random(size=(N, 200))
+            self.y = np.random.randint(2, size=(N,))
+        else:
+            raise ValueError("Unknown dataset name: {}".format(dataset_name))
+
+        offset = int(self.X.shape[0] * (1 - test_fraction))
+        self.X_train, self.y_train = self.X[:offset], self.y[:offset]
+        self.X_test, self.y_test = self.X[offset:], self.y[offset:]
+
+    @classmethod
+    def get_instance(cls, dataset_name, test_fraction=0.02):
+        key = dataset_name + " {}".format(test_fraction)
```

Comment:
This key bothers me a little. Shouldn't it be unique per dataset, fraction and a model? The model part seems to be missing here.

Reply:
Oops, sorry, taking this back. Model is passed later. Please ignore.